Documentation/login-spec.md

```diff
@@ -0,0 +1,302 @@
+# WordPress Login Specification
+
+This document specifies the requirements and behaviors for WordPress.org site login functionality. The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+# 1: Standard SSL-Enabled WordPress Installation
+
+Requirements:
+1. The site MUST be a stock WordPress Installation without modifications.
+2. The site MUST use SSL.
+3. The site MUST NOT have a CDN or caching that prevents the `Link` header from being sent.
+
+Expected Signals:
+1. The `Link` header MUST be present
+2. The `authorization_url` MUST be present in the JSON API root
+
+Behavior:
+1. The system MUST complete login successfully.
+2. The system MUST NOT display any warnings.
+
+Reference Implementation: https://vanilla.wpmt.co
+
+# 2: Local Development Environment
+
+Requirements:
+1. The site MUST be a stock WordPress Installation.
+2. The site MAY have an invalid SSL certificate.
+3. The `WP_ENVIRONMENT_TYPE` MUST NOT be set to "production"
+
+Expected Signals:
+1. The `Link` header MUST be present
+2. The `authorization_url` MUST NOT be present in the JSON API root
+3. The `has_https` check MUST fail
+4. The URL MUST match one of:
+   - An IP address
+   - A domain ending in `.dev`
+   - A domain ending in `.local`
+   - The hostname `localhost`
+
+Behavior:
+1. The system MUST reject the login attempt
+2. The system MUST display the error: "This site is a local development environment. You'll need to enable application passwords to connect to it with the app."
+
+Reference Implementation: http://localhost
+
+# 3: WordPress Admin URL Entry
```

Comment:
This says `WordPress Admin`, but it handles both `/wp-login` and `wp-admin` cases, should we rename it?